File: frontend/services/api/jobs.ts
import { fetchJson } from "@/services/api/client";
import { ApiListResponse } from "@/types/api";
import { JobCreateInput, JobModel, JobStatus } from "@/types/job";

export async function submitJob(payload: JobCreateInput): Promise<JobModel> {
  return fetchJson<JobModel>("/api/v1/jobs", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getJob(jobId: string): Promise<JobModel> {
  return fetchJson<JobModel>(`/api/v1/jobs/${jobId}`, {
    retries: 1,
    backoffMs: 250,
  });
}

export async function listJobs(status?: JobStatus): Promise<JobModel[]> {
  const suffix = status ? `?status=${status}` : "";
  const response = await fetchJson<ApiListResponse<JobModel>>(`/api/v1/jobs${suffix}`);
  return response.items;
}

export async function retryJob(jobId: string): Promise<JobModel> {
  return fetchJson<JobModel>(`/api/v1/jobs/${jobId}/retry`, {
    method: "POST",
  });
}

File: backend/app/api/routes/jobs.py
from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.api.deps import get_orchestrator, get_store
from app.models.job import Job, JobCreateRequest, JobListResponse, JobStatus
from app.services.orchestrator import JobOrchestrator
from app.services.state_store import InMemoryStateStore

router = APIRouter(prefix="/jobs", tags=["jobs"])


@router.post("", response_model=Job, status_code=status.HTTP_202_ACCEPTED)
async def submit_job(
    payload: JobCreateRequest,
    orchestrator: JobOrchestrator = Depends(get_orchestrator),
) -> Job:
    try:
        return await orchestrator.submit_job(payload)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_402_PAYMENT_REQUIRED, detail=str(exc)) from exc


@router.get("/{job_id}", response_model=Job)
async def get_job(job_id: str, store: InMemoryStateStore = Depends(get_store)) -> Job:
    job = await store.get_job(job_id)
    if not job:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job not found")
    return job


@router.get("", response_model=JobListResponse)
async def list_jobs(
    status_filter: JobStatus | None = Query(default=None, alias="status"),
    store: InMemoryStateStore = Depends(get_store),
) -> JobListResponse:
    jobs = await store.list_jobs()
    if status_filter:
        jobs = [job for job in jobs if job.status == status_filter]
    return JobListResponse(items=jobs)


@router.post("/{job_id}/retry", response_model=Job, status_code=status.HTTP_202_ACCEPTED)
async def retry_job(
    job_id: str,
    orchestrator: JobOrchestrator = Depends(get_orchestrator),
) -> Job:
    try:
        return await orchestrator.retry_job(job_id)
    except KeyError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_402_PAYMENT_REQUIRED, detail=str(exc)) from exc
